Fix: Solve bug in intersect_reals

shashankks0987 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #21595


#### Brief description of what is fixed or changed
earlier, imageset(Lambda(n, n + (n - 1)*(n + 1)*I), S.Integers).intersect(S.Reals) returned an incorrect result
now, the solution is FiniteSet(1,1) which is the correct solution

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* sets
   * Fixed a bug in is_subset()
<!-- END RELEASE NOTES -->

shashankks0987 committed:
Fix: Solve bug in intersect_reals
earlier, imageset(Lambda(n, n + (n - 1)*(n + 1)*I), S.Integers).intersect(S.Reals) returned an incorrect result
now, the solution is FiniteSet(1,1) which is the correct solution

shashankks0987 committed:
Remove unused imports

oscarbenjamin commented on sympy/sets/handlers/intersection.py:
```diff
@@ -306,7 +306,7 @@ def intersection_sets(self, other): # noqa:F811
             x, xis = zip(*[solve_linear(i, 0) for i in Mul.make_args(im) if n in i.free_symbols])
             if x and all(i == n for i in x):
-                base_set -= FiniteSet(xis)
+                base_set = FiniteSet(*xis)
```
Should it be `base_set &= FiniteSet(*xis)`?

smichr replied:
If the imaginary part is never 0 then there is no portion of the base set that will ever be real so the empty set is returned. Otherwise wherever the imaginary portion of the base set is 0 -- *within the base set* -- is where the intersection will be real. So I think you are right, @oscarbenjamin . e.g. in the test that was changed, if the expression were changed to `(n - 1)*(n - S.Half)*I` then `xis` is `{1, 1/2}` but only 1 is in the base set of Integers, so the answer should be {1}.

smichr commented on sympy/sets/tests/test_fancysets.py:
```diff
@@ -657,7 +657,7 @@ def test_infinitely_indexed_set_2():
     from sympy import I
     from sympy.abc import n
-    assert imageset(Lambda(n, n + (n - 1)*(n + 1)*I), S.Integers).intersect(S.Reals) == Complement(S.Integers, FiniteSet((-1, 1)))
+    assert imageset(Lambda(n, n + (n - 1)*(n + 1)*I), S.Integers).intersect(S.Reals) == FiniteSet(-1, 1)
```
```suggestion
    assert imageset(Lambda(n, n + (n - 1)*(n + S.Half)*I), S.Integers).intersect(S.Reals) == FiniteSet(1)
```

smichr commented on sympy/sets/handlers/intersection.py:
```diff
@@ -306,7 +306,7 @@ def intersection_sets(self, other): # noqa:F811
-                base_set -= FiniteSet(xis)
+                base_set = FiniteSet(*xis)
             else:
                 base_set -= ConditionSet(n, Eq(im, 0), S.Integers)
```
```suggestion
                base_set = ConditionSet(n, Eq(im, 0), base_set)
```

smichr replied:
I suspect the same change should be made below, too, at `base_set -= ConditionSet(n, Eq(i, 0), S.Integers)`, changing `S.Integers` to `base_set`. And in these cases `-=`, not `&=` is correct, right, because we are removing values of n that are invalid.

shashankks0987 committed:
Fix: Implementation of intersection_sets
If the imaginary part is never 0 then there is no portion of the base set that will ever be real so an empty set is returned. Otherwise wherever the imaginary portion of the base set is 0 -- within the base set --  is where the intersection will be real.

shashankks0987 commented on sympy/sets/handlers/intersection.py:
```diff
@@ -306,17 +306,17 @@ def intersection_sets(self, other): # noqa:F811
-                        base_set -= FiniteSet(xis)
+                        base_set &= FiniteSet(*xis)
                 else:
                     base_set -= ConditionSet(n, Eq(i, 0), S.Integers)
```
Does this line need to change too? @smichr

shashankks0987 committed:
Fix: excluding values that make denominators 0 in intersect_sets
- earlier the values were added to the base set instead of removing them

smichr commented on sympy/sets/handlers/intersection.py:
```diff
@@ -306,17 +306,17 @@ def intersection_sets(self, other): # noqa:F811
                     x, xis = sol
                     if x and all(i == n for i in x):
-                        base_set -= FiniteSet(xis)
+                        base_set &= FiniteSet(*xis)
```
This one was correct -- we must exclude from the base set any solutions of denoms. Check what you get for `Lambda(n, n/3 + (n - 2)*(n - 1)*I/(2*n - 2), Integers).intersection(Reals)` -- the answer should be `{2/3}`

shashankks0987 replied:
ImageSet(Lambda(x, x/2), ConditionSet(n, Eq((n - 2)*(n - 1)/(2*n - 2), 0), Integers)) I get this with the &=

shashankks0987 replied:
```
In [1]: s1=imageset(Lambda(n, n/2 + (n - 2)*(n - 1)*I/(2*n - 2)), Integers)

In [2]: s1.intersect(Reals)
Out[2]: 
⎧x │     ⎧  │         ⎛(n - 2)⋅(n - 1)    ⎞⎫⎫
⎨─ │ x ∊ ⎨n │ n ∊ ℤ ∧ ⎜─────────────── = 0⎟⎬⎬
⎩2 │     ⎩  │         ⎝    2⋅n - 2        ⎠⎭⎭

```

smichr commented:
I need some time to check this out. Please give me a day.

smichr commented:
```diff
diff --git a/sympy/sets/handlers/intersection.py b/sympy/sets/handlers/intersection.py
index d40ed94b7c..660bc8fa1a 100644
--- a/sympy/sets/handlers/intersection.py
+++ b/sympy/sets/handlers/intersection.py
@@ -5,6 +5,7 @@
 from sympy.sets.fancysets import (Integers, Naturals, Reals, Range,
     ImageSet, Rationals)
 from sympy.sets.sets import UniversalSet, imageset, ProductSet
+from sympy.simplify.radsimp import numer
 
 
 @dispatch(ConditionSet, ConditionSet)  # type: ignore # noqa:F811
@@ -304,21 +305,24 @@ def intersection_sets(self, other): # noqa:F811
             return None
         else:
             # univarite imaginary part in same variable
-            x, xis = zip(*[solve_linear(i, 0) for i in Mul.make_args(im) if n in i.free_symbols])
-            if x and all(i == n for i in x):
-                base_set &= FiniteSet(*xis)
-            else:
-                base_set = ConditionSet(n, Eq(im, 0), base_set)
+            im0 = S.EmptySet
+            # use numer instead of as_numer_denom to keep
+            # this as fast as possible while still handling
+            # simple cases
+            for i in Mul.make_args(numer(im)):
+                x, xis = solve_linear(i, 0, [n])
+                if x == n:
+                    im0 |= FiniteSet(xis) & base_set
+                else:
+                    im0 |= ConditionSet(n, Eq(i, 0), base_set)
+            base_set = im0
         # exclude values that make denominators 0
         for i in denoms(f):
-            if i.has(n):
-                sol = list(zip(*[solve_linear(i, 0) for i in Mul.make_args(im) if n in i.free_symbols]))
-                if sol != []:
-                    x, xis = sol
-                    if x and all(i == n for i in x):
-                        base_set -= FiniteSet(*xis)
-                else:
-                    base_set -= ConditionSet(n, Eq(i, 0), base_set)
+            x, xis = solve_linear(i, 0, [n])
+            if x == n:
+                base_set -= FiniteSet(xis)
+            else:
+                base_set -= ConditionSet(n, Eq(i, 0), base_set)
         return imageset(lam, base_set)
 
     elif isinstance(other, Interval):
diff --git a/sympy/sets/tests/test_fancysets.py b/sympy/sets/tests/test_fancysets.py
index 30ad17cffe..da780c81b1 100644
--- a/sympy/sets/tests/test_fancysets.py
+++ b/sympy/sets/tests/test_fancysets.py
@@ -1,9 +1,11 @@
 
 from sympy.core.expr import unchanged
+from sympy.sets.conditionset import ConditionSet
 from sympy.sets.fancysets import (ImageSet, Range, normalize_theta_set,
                                   ComplexRegion)
 from sympy.sets.sets import (FiniteSet, Interval, Union, imageset,
-                             Intersection, ProductSet, Contains)
+                             Intersection, ProductSet, Contains,
+                             Complement)
 from sympy.simplify.simplify import simplify
 from sympy import (S, Symbol, Lambda, symbols, cos, sin, pi, oo, Basic,
                    Rational, sqrt, tan, log, exp, Abs, I, Tuple, eye,
@@ -657,8 +659,22 @@ def test_infinitely_indexed_set_2():
 def test_imageset_intersect_real():
     from sympy import I
     from sympy.abc import n
-    assert imageset(Lambda(n, n + (n - 1)*(n + 1)*I), S.Integers).intersect(S.Reals) == FiniteSet(-1, 1)
-    assert imageset(Lambda(n, n + (n - 1)*(n + S.Half)*I), S.Integers).intersect(S.Reals) == FiniteSet(1)
+    im = (n - 1)*(n + S.Half)
+    assert imageset(Lambda(n, n + im*I), S.Integers
+        ).intersect(S.Reals) == FiniteSet(1)
+    assert imageset(Lambda(n, n + im*(n + 1)*I), S.Naturals0
+        ).intersect(S.Reals) == FiniteSet(1)
+    assert imageset(Lambda(n, n/2 + im.expand()*I), S.Integers
+        ).intersect(S.Reals) == ImageSet(Lambda(x, x/2), ConditionSet(
+        n, Eq(n**2 - n/2 - S(1)/2, 0), S.Integers))
+    assert imageset(Lambda(n, n/(1/n - 1) + im*(n + 1)*I), S.Integers
+        ).intersect(S.Reals) == FiniteSet(S.Half)
+    assert imageset(Lambda(n, n/(n - 6) +
+        (n - 3)*(n + 1)*I/(2*n + 2)), S.Integers).intersect(S.Reals
+        ) == FiniteSet(-1)
+    assert imageset(Lambda(n, n/(n**2 - 9) +
+        (n - 3)*(n + 1)*I/(2*n + 2)), S.Integers).intersect(S.Reals
+        ) is S.EmptySet
     s = ImageSet(
         Lambda(n, -I*(I*(2*pi*n - pi/4) + log(Abs(sqrt(-I))))),
         S.Integers)
```

shashankks0987 committed:
Fix: Implementation of intersection_sets to use numer
- Add tests

smichr commented on sympy/sets/tests/test_fancysets.py:
```diff
@@ -657,7 +658,19 @@ def test_infinitely_indexed_set_2():
+    im = (n - 1)*(n + S.Half)
+    assert imageset(Lambda(n, n + im*I), S.Integers
+        ).intersect(S.Reals) == FiniteSet(1)
+    assert imageset(Lambda(n, n + im*(n + 1)*I), S.Naturals0).intersect(S.Reals) == FiniteSet(1)
+    assert imageset(Lambda(n, n/2 + im.expand()*I), S.Integers
+).intersect(S.Reals) == ImageSet(Lambda(x, x/2), ConditionSet(
+        n, Eq(n**2 - n/2 - S(1)/2, 0), S.Integers))
+    assert imageset(Lambda(n, n/(1/n - 1) + im*(n + 1)*I), S.Integers).intersect(S.Reals) == FiniteSet(S.Half)
+    assert imageset(Lambda(n, n/(n - 6) +
+        (n - 3)*(n + 1)*I/(2*n + 2)), S.Integers).intersect(S.Reals) == FiniteSet(-1)
+    assert imageset(Lambda(n, n/(n**2 - 9) +
+        (n - 3)*(n + 1)*I/(2*n + 2)), S.Integers).intersect(S.Reals) is S.EmptySet
```
```suggestion
    assert imageset(Lambda(n, n + im*(n + 1)*I), S.Naturals0
        ).intersect(S.Reals) == FiniteSet(1)
    assert imageset(Lambda(n, n/2 + im.expand()*I), S.Integers
        ).intersect(S.Reals) == ImageSet(Lambda(x, x/2), ConditionSet(
        n, Eq(n**2 - n/2 - S(1)/2, 0), S.Integers))
    assert imageset(Lambda(n, n/(1/n - 1) + im*(n + 1)*I), S.Integers
        ).intersect(S.Reals) == FiniteSet(S.Half)
    assert imageset(Lambda(n, n/(n - 6) +
        (n - 3)*(n + 1)*I/(2*n + 2)), S.Integers).intersect(
        S.Reals) == FiniteSet(-1)
    assert imageset(Lambda(n, n/(n**2 - 9) +
        (n - 3)*(n + 1)*I/(2*n + 2)), S.Integers).intersect(
        S.Reals) is S.EmptySet
```

smichr committed:
Update sympy/sets/tests/test_fancysets.py

smichr commented:
Do you want to give this a final check, @oscarbenjamin ?

oscarbenjamin commented on sympy/sets/handlers/intersection.py:
```diff
@@ -304,21 +304,24 @@ def intersection_sets(self, other): # noqa:F811
+            for i in Mul.make_args(numer(im)):
+                x, xis = solve_linear(i, 0, [n])
+                if x == n:
+                    sol_set |= FiniteSet(xis) & base_set
```
Building a set with `|` in a loop leads to quadratic behaviour e.g.:
```python
In [1]: sets = imageset(lambda n: I*prod(n-i for i in range(200)), Integers)

In [2]: %time sets & Reals
CPU times: user 1.31 s, sys: 5.47 ms, total: 1.32 s
Wall time: 1.32 s
Out[2]: {0}

In [3]: sets = imageset(lambda n: I*prod(n-i for i in range(400)), Integers)

In [4]: %time sets & Reals
CPU times: user 4.43 s, sys: 9.75 ms, total: 4.44 s
Wall time: 4.44 s
Out[4]: {0}
```

oscarbenjamin commented:
I don't object to this PR but performance should be checked carefully.

We badly need to move any complicated logic like this out of these handlers and into some kind of non-automatic simplification routines. The problem with having them in the handlers is that in the case where they can't evaluate they still run every time any operation like `subs` rebuilds the tree. So every operation with an `Intersection` is slowed down just to add a bit of code to handle some potentially obscure case.

smichr committed:
Update intersection.py

smichr commented:
This is less handling now than was done previously. Leaving extraction of simplification code and determination of where to put it to another issue.
